bika/lims/validators.py

```diff
@@ -77,73 +74,119 @@ def __call__(self, value, *args, **kwargs):
+                        "to any catalog!".format(repr(context)))
+            return None
+
+        # take the first catalog
         catalog = catalogs[0]
 
-        # 2. Check if the field accessor is indexed
-        field_index = None
-        accessor = field.getAccessor(instance)
+        # Check if the field accessor is indexed
+        field_index = field.getName()
+        accessor = field.getAccessor(context)
         if accessor:
             field_index = accessor.__name__
 
-        # 3. Check if the field index is in the indexes
-        # Field is indexed, use the catalog instead of objectValues
-        parent_path = api.get_parent_path(instance)
-        portal_type = instance.portal_type
-        catalog_query = {"portal_type": portal_type,
-                         "path": {"query": parent_path, "depth": 1}}
-
-        # We try here to avoid waking up all the objects, because this can be
-        # likely very expensive if the parent object contains many objects
-        if fieldname in catalog.indexes():
-            # We use the fieldname as index to reduce the results list
-            catalog_query[fieldname] = to_utf8(safe_unicode(value))
-            parent_objects = map(api.get_object, catalog(catalog_query))
-        elif field_index and field_index in catalog.indexes():
-            # We use the field index to reduce the results list
-            catalog_query[field_index] = to_utf8(safe_unicode(value))
-            parent_objects = map(api.get_object, catalog(catalog_query))
-        else:
-            # fall back to the objectValues :(
-            parent_object = api.get_parent(instance)
-            parent_objects = parent_object.objectValues()
+        # return if the field is not indexed
+        if field_index not in catalog.indexes():
+            return None
+
+        # build a catalog query
+        query = {
+            "portal_type": api.get_portal_type(context),
+            "path": {
+                "query": api.get_parent_path(context),
+                "depth": 1,
+            }
+        }
+        query[field_index] = value
+        logger.info("UniqueFieldValidator:Query={}".format(query))
+        return query
+
+    def __call__(self, value, *args, **kwargs):
+        context = kwargs['instance']
+        uid = api.get_uid(context)
+        field = kwargs['field']
+        fieldname = field.getName()
+        translate = getToolByName(context, 'translation_service').translate
+
+        # return directly if nothing changed
+        if value == field.get(context):
+            return True
+
+        # Fetch the parent object candidates by catalog or by objectValues
+        #
+        # N.B. We want to use the catalog to speed things up, because using
+        # `parent.objectValues` is very expensive if the parent object contains
+        # many items and causes the UI to block too long
+        catalog_query = self.make_catalog_query(context, field, value)
```

Comment:
If I am not wrong, searches against `FieldIndex` type indexes only return values that matches entirely with the search term (http://zope.readthedocs.io/en/latest/zope2book/SearchingZCatalog.html#searching-field-indexes), so I if the search term is "Foo", it will not return those brains with values "Foo-1" or "Foo-2" for the same index. But understand that this would mean an extra check to be sure the type of the index is `FieldIndex`. Worth to mention that another solution would be to use the `uniqueValues()` function, that retrieves all unique values for a certain index: https://docs.plone.org/develop/plone/searching_and_indexing/query.html#unique-values

Re successful catalog query, you are pretty right here, but well.. if we have to take into account the eventuality of a non indexed field everywhere, then "Houston, we have a problem". Re `UnicodeDecodeError` (as well as other errors that may rise):

```python
catalog_query = self.make_catalog_query(context, field, value)
if catalog_query:
    try:
        catalog = api.get_catalogs_for(context)[0]
        brains = catalog(catalog_query)
        return len(brains) == 0
    except (IndexError, UnicodeDecodeError, ParseError, api.BikaLIMSError) as e:
        logger.warn('Something really wrong happened!')
parent_objects = self.get_parent_objects(context)
[...]
```
Of course, would be better if all this block of code was encapsulated in another function, or even reuse `make_catalog_query` by renaming it and let it make the query and return the brains if all is ok or `None` otherwise. An extra-bonus is that then, you'll not need to call `api.get_catalogs_for(context)` twice:

```python
brains = self.try_catalog_query(context, field, value)
if brains:
    return len(brains) == 0
parent_objects = self.get_parent_objects(context)
[...]
```